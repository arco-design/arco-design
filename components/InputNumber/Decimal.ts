import {
  toSafeString,
  trimNumber,
  validateNumber,
  getNumberPrecision,
  supportBigInt,
} from './utils';

export interface Decimal {
  readonly isEmpty: boolean;
  readonly isNaN: boolean;
  readonly isInvalid: boolean;
  toNumber: () => number;
  toString: (option?: { safe: boolean; precision?: number }) => string;
  equals: (target: Decimal) => boolean;
  less: (target: Decimal) => boolean;
  negate: () => Decimal;
  add: (target: string | number) => Decimal;
}

class BigIntDecimal implements Decimal {
  readonly isEmpty: boolean;

  readonly isNaN: boolean;

  private readonly isNegative: boolean;

  private readonly origin: string = '';

  private readonly integer: bigint;

  private readonly decimal: bigint;

  private readonly decimalLen: number;

  constructor(value: string | number) {
    this.origin = String(value);

    if ((!value && value !== 0) || !this.origin.trim()) {
      this.isEmpty = true;
      return;
    }

    if (value === '-') {
      this.isNaN = true;
      return;
    }

    const safeValueString = toSafeString(value);
    if (validateNumber(safeValueString)) {
      const { negative, trimStr } = trimNumber(safeValueString);
      const [integerStr, decimalStr = '0'] = trimStr.split('.');
      this.isNegative = negative;
      this.integer = BigInt(integerStr);
      this.decimal = BigInt(decimalStr);
      this.decimalLen = decimalStr.length;
    } else {
      this.isNaN = true;
    }
  }

  get isInvalid() {
    return this.isEmpty || this.isNaN;
  }

  private getMark() {
    return this.isNegative ? '-' : '';
  }

  private getIntegerStr() {
    return this.integer.toString();
  }

  private getDecimalStr() {
    return this.decimal.toString().padStart(this.decimalLen, '0');
  }

  private alignDecimal(decimalLength: number): bigint {
    return BigInt(
      `${this.getMark()}${this.getIntegerStr()}${this.getDecimalStr().padEnd(decimalLength, '0')}`
    );
  }

  negate() {
    const numStr = this.toString();
    return new BigIntDecimal(numStr.startsWith('-') ? numStr.slice(1) : `-${numStr}`);
  }

  add(value: string | number): BigIntDecimal {
    const offset = new BigIntDecimal(value);

    if (offset.isInvalid) {
      return this;
    }

    if (this.isInvalid) {
      return offset;
    }

    const maxDecimalLength = Math.max(this.decimalLen, offset.decimalLen);
    const thisAlignedDecimal = this.alignDecimal(maxDecimalLength);
    const offsetAlignedDecimal = offset.alignDecimal(maxDecimalLength);
    const valueStr = (thisAlignedDecimal + offsetAlignedDecimal).toString();
    const { negativeStr, trimStr } = trimNumber(valueStr);
    const hydrateValueStr = `${negativeStr}${trimStr.padStart(maxDecimalLength + 1, '0')}`;

    return new BigIntDecimal(
      `${hydrateValueStr.slice(0, -maxDecimalLength)}.${hydrateValueStr.slice(-maxDecimalLength)}`
    );
  }

  equals(target: BigIntDecimal) {
    return this.toString() === target?.toString();
  }

  less(target: BigIntDecimal) {
    return this.isInvalid || target.isInvalid
      ? false
      : this.add(target.negate().toString()).toNumber() < 0;
  }

  toNumber(): number {
    return this.isNaN ? NaN : Number(this.toString());
  }

  toString(options: { safe: boolean; precision?: number } = { safe: true }): string {
    const { safe, precision } = options;
    const result = safe
      ? this.isInvalid
        ? ''
        : trimNumber(`${this.getMark()}${this.getIntegerStr()}.${this.getDecimalStr()}`).fullStr
      : this.origin;
    return typeof precision === 'number' ? toFixed(result, precision) : result;
  }
}

class NumberDecimal implements Decimal {
  readonly isEmpty: boolean;

  readonly isNaN: boolean;

  private readonly origin: string = '';

  private readonly number: number;

  constructor(value: string | number) {
    this.origin = String(value);
    this.number = Number(value);

    if ((!value && value !== 0) || !this.origin.trim()) {
      this.isEmpty = true;
    } else {
      this.isNaN = Number.isNaN(this.number);
    }
  }

  get isInvalid() {
    return this.isEmpty || this.isNaN;
  }

  negate() {
    return new NumberDecimal(-this.toNumber());
  }

  equals(target: NumberDecimal) {
    return this.toNumber() === target?.toNumber();
  }

  less(target: NumberDecimal) {
    return this.isInvalid || target.isInvalid
      ? false
      : this.add(target.negate().toString()).toNumber() < 0;
  }

  add(value: string | number): NumberDecimal {
    const offset = new NumberDecimal(value);

    if (offset.isInvalid) {
      return this;
    }

    if (this.isInvalid) {
      return offset;
    }

    const result = this.number + offset.number;
    if (result > Number.MAX_SAFE_INTEGER) {
      return new NumberDecimal(Number.MAX_SAFE_INTEGER);
    }

    if (result < Number.MIN_SAFE_INTEGER) {
      return new NumberDecimal(Number.MIN_SAFE_INTEGER);
    }

    const maxPrecision = Math.max(
      getNumberPrecision(this.number),
      getNumberPrecision(offset.number)
    );
    return new NumberDecimal(result.toFixed(maxPrecision));
  }

  toNumber() {
    return this.number;
  }

  toString(options: { safe: boolean; precision?: number } = { safe: true }) {
    const { safe, precision } = options;
    const result = safe ? (this.isInvalid ? '' : toSafeString(this.number)) : this.origin;
    return typeof precision === 'number' ? toFixed(result, precision) : result;
  }
}

export function getDecimal(value: string | number): Decimal {
  return supportBigInt() ? new BigIntDecimal(value) : new NumberDecimal(value);
}

/**
 * Replace String.prototype.toFixed like Math.round
 * If cutOnly is true, just slice the tail
 * e.g. Decimal.toFixed(0.15) will return 0.2, not 0.1
 */
export function toFixed(numStr: string, precision?: number, cutOnly = false): string {
  if (numStr === '') {
    return '';
  }

  const separator = '.';
  const { negativeStr, integerStr, decimalStr } = trimNumber(numStr);
  const precisionDecimalStr = `${separator}${decimalStr}`;
  const numberWithoutDecimal = `${negativeStr}${integerStr}`;

  if (precision >= 0) {
    const advancedNum = Number(decimalStr[precision]);
    if (advancedNum >= 5 && !cutOnly) {
      const advancedDecimal = getDecimal(numStr).add(
        `${negativeStr}0.${'0'.repeat(precision)}${10 - advancedNum}`
      );
      return toFixed(advancedDecimal.toString(), precision, cutOnly);
    }

    return precision === 0
      ? numberWithoutDecimal
      : `${numberWithoutDecimal}${separator}${decimalStr
          .padEnd(precision, '0')
          .slice(0, precision)}`;
  }

  return `${numberWithoutDecimal}${precisionDecimalStr === '.0' ? '' : precisionDecimalStr}`;
}
