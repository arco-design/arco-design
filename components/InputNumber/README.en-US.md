`````
Component / Data Entry

# InputNumber

An input box which only allows to enter number.
`````

%%Content%%

## API

### InputNumber

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|disabled|Whether input is disabled|boolean |`-`|-|
|error|Whether input has error status|boolean |`-`|-|
|hideControl|Whether to hide the control buttons|boolean |`-`|-|
|readOnly|Whether input is readonly|boolean |`-`|2.17.0|
|strictMode|`onChange` will return a string in strict mode|boolean |`-`|2.42.0|
|max|The max value|number |`Infinity`|-|
|min|The min value|number |`-Infinity`|-|
|precision|The precision of input value|number |`-`|-|
|step|The value of each step. It can be an integer or decimal|number |`1`|-|
|placeholder|Placeholder of input element|string |`-`|-|
|mode|`embed`: Button embedded, `button`: Buttons on both sides|'embed' \| 'button' |`embed`|-|
|size|Different sizes|'mini' \| 'small' \| 'default' \| 'large' |`-`|-|
|prefix|The prefix for the InputNumber|ReactNode |`-`|-|
|suffix|The suffix for the InputNumber|ReactNode |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|defaultValue|To set default value|number \| string |`-`|-|
|icons|Customize icons|{up?: ReactNode;down?: ReactNode;plus?: ReactNode;minus?: ReactNode;} |`-`|-|
|style|Additional style|CSSProperties |`-`|-|
|value|To set value|undefined \| number \| string |`-`|-|
|formatter|Specifies the format of the value presented|(value: number \| string, info: { userTyping: boolean; input: string }) => string |`-`|Param `info` in `2.41.0`|
|onBlur|Callback when the input is blurred|(e) => void |`-`|-|
|onChange|Callback when the value changes|(value: number) => void |`-`|-|
|onFocus|Callback when the input is focused|(e) => void |`-`|-|
|onKeyDown|Callback when the keyboard is pressed|(e: Event) => void |`-`|-|
|parser|Specifies the value extracted from formatter|(value: string) => number \| string |`(input) => input.replace(/[^\w\.-]+/g, '')`|-|

## Methods

|Name|Description|
|---|:---:|
|blur()|Remove focus|
|focus()|Get focus|
