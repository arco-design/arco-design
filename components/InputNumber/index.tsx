import React, {
  useCallback,
  useContext,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
} from 'react';
import NP from 'number-precision';
import IconUp from '../../icon/react-icon/IconUp';
import IconDown from '../../icon/react-icon/IconDown';
import IconPlus from '../../icon/react-icon/IconPlus';
import IconMinus from '../../icon/react-icon/IconMinus';
import { isNumber } from '../_util/is';
import cs from '../_util/classNames';
import { ArrowUp, ArrowDown } from '../_util/keycode';
import { ConfigContext } from '../ConfigProvider';
import Input from '../Input';
import { RefInputType } from '../Input/interface';
import { InputNumberProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';
import omit from '../_util/omit';

NP.enableBoundaryChecking(false);

// Value's auto change speed when user holds on plus or minus
const AUTO_CHANGE_INTERVAL = 200;

// Delay to auto change value when user holds on plus or minus
const AUTO_CHANGE_START_DELAY = 1000;

type StepMethods = 'minus' | 'plus';

const defaultProps: InputNumberProps = {
  max: Infinity,
  min: -Infinity,
  step: 1,
  mode: 'embed',
  parser: (input) => input.replace(/[^\w\.-]+/g, ''),
};

function InputNumber(baseProps: InputNumberProps, ref) {
  const { getPrefixCls, size: ctxSize, componentConfig } = useContext(ConfigContext);
  const props = useMergeProps<InputNumberProps>(
    baseProps,
    defaultProps,
    componentConfig?.InputNumber
  );
  const {
    className,
    style,
    defaultValue,
    disabled,
    error,
    readOnly,
    placeholder,
    hideControl,
    suffix,
    prefix,
    icons,
    mode,
    size,
    step,
    precision,
    min,
    max,
    parser,
    formatter,
    onBlur,
    onFocus,
    onChange,
    onKeyDown,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('input-number');
  const mergedSize = size || ctxSize;
  const mergedPrecision = (() => {
    if (isNumber(precision)) {
      const decimal = `${step}`.split('.')[1];
      const stepPrecision = (decimal && decimal.length) || 0;
      return Math.max(stepPrecision, precision);
    }
    return null;
  })();

  const [innerValue, setInnerValue] = useState<InputNumberProps['value']>(
    'defaultValue' in props ? defaultValue : undefined
  );
  const value = (() => {
    const mergedValue = 'value' in props ? props.value : innerValue;
    return typeof mergedValue === 'string' && mergedValue !== '' ? +mergedValue : mergedValue;
  })();

  const [inputValue, setInputValue] = useState<string>('');
  const [isOutOfRange, setIsOutOfRange] = useState(false);
  const [isUserInputting, setIsUserInputting] = useState(false);

  // Value is not set
  const isEmptyValue = value === '' || value === undefined || value === null;

  const refAutoTimer = useRef(null);
  const refInput = useRef<RefInputType>(null);
  // Ref to keep track of whether user has taken operations since the last change of prop value
  const refHasOperateSincePropValueChanged = useRef(false);

  useImperativeHandle(ref, () => refInput.current, []);

  const setValue = (newVal) => {
    setInnerValue(newVal);

    const newValue = isNumber(+newVal) ? +newVal : undefined;
    if (newValue !== value) {
      onChange && onChange(newValue);
    }
  };

  const stop = () => {
    refAutoTimer.current && clearTimeout(refAutoTimer.current);
    refAutoTimer.current = null;
  };

  const getLegalValue = useCallback(
    (changedValue) => {
      let finalValue: string | number = Number(changedValue);

      if (!changedValue && changedValue !== 0) {
        finalValue = undefined;
      } else if (!isNumber(finalValue)) {
        finalValue = changedValue === '-' ? changedValue : '';
      }

      if (finalValue < min) {
        finalValue = min;
      }

      if (finalValue > max) {
        finalValue = max;
      }

      return isNumber(finalValue)
        ? isNumber(mergedPrecision)
          ? Number(finalValue.toFixed(mergedPrecision))
          : finalValue
        : undefined;
    },
    [min, max, mergedPrecision]
  );

  useEffect(() => {
    return () => stop();
  }, []);

  useEffect(() => {
    refHasOperateSincePropValueChanged.current = false;
  }, [props.value]);

  useEffect(() => {
    const _isOutOfRange = (isNumber(min) && value < min) || (isNumber(max) && value > max);

    // Don't correct the illegal value caused by prop value. Wait for user to take actions.
    if (_isOutOfRange && refHasOperateSincePropValueChanged.current) {
      setValue(getLegalValue(value));
    }

    setIsOutOfRange(_isOutOfRange);
  }, [min, max, value, getLegalValue]);

  const handleArrowKey = (event, method: StepMethods, needRepeat = false) => {
    event.persist();
    event.preventDefault();
    setIsUserInputting(false);

    if (disabled) {
      return;
    }

    let finalValue = min === -Infinity ? 0 : min;

    if (!isEmptyValue) {
      finalValue = NP[method](value, step);
    }

    setValue(getLegalValue(finalValue));
    refInput.current && refInput.current.focus();

    // auto change while holding
    if (needRepeat) {
      const isFirstRepeat = refAutoTimer.current === null;
      refAutoTimer.current = setTimeout(
        () => event.target.dispatchEvent(event.nativeEvent),
        isFirstRepeat ? AUTO_CHANGE_START_DELAY : AUTO_CHANGE_INTERVAL
      );
    }
  };

  const getDisplayInputValue = () => {
    let _value: string | number;

    if (isUserInputting) {
      _value = inputValue;
    } else if (isNumber(value) && isNumber(mergedPrecision)) {
      _value = value.toFixed(mergedPrecision);
    } else if (value == null) {
      _value = '';
    } else {
      _value = value.toString();
    }

    return formatter ? formatter(_value) : _value;
  };

  const inputEventHandlers = {
    onChange: (value) => {
      setIsUserInputting(true);

      let targetValue = value.trim().replace(/ã€‚/g, '.');
      targetValue = parser ? parser(targetValue) : targetValue;

      if (isNumber(+targetValue) || targetValue === '-' || !targetValue || targetValue === '.') {
        const formatValue = getLegalValue(targetValue);

        setInputValue(targetValue);
        setValue(formatValue);
      }
    },
    onKeyDown: (e) => {
      e.stopPropagation();

      const key = e.key;
      if (key === ArrowDown.key) {
        handleArrowKey(e, 'minus');
      } else if (key === ArrowUp.key) {
        handleArrowKey(e, 'plus');
      }

      stop();
      onKeyDown && onKeyDown(e);
    },
    onFocus: (e) => {
      // Both tab and button click trigger focus event. This can be used to determine whether user has taken operations
      refHasOperateSincePropValueChanged.current = true;
      setInputValue(refInput.current?.dom?.value);
      onFocus && onFocus(e);
    },
    onBlur: (e) => {
      setValue(getLegalValue(value));
      setIsUserInputting(false);
      onBlur && onBlur(e);
    },
  };

  const getControlButtonEventsHandlers = (method: StepMethods) => {
    return readOnly
      ? {}
      : {
          onMouseDown: (e) => handleArrowKey(e, method, true),
          onMouseLeave: stop,
          onMouseUp: stop,
        };
  };

  const shouldRenderButton = !hideControl && mode === 'button';
  const shouldRenderLayer = !hideControl && !readOnly && mode === 'embed';

  const renderStepButton = (method: StepMethods, icon) => {
    return (
      <div
        className={cs(`${prefixCls}-step-button`, {
          [`${prefixCls}-step-button-disabled`]:
            disabled || (method === 'plus' ? +value >= +max : +value <= +min),
        })}
        {...getControlButtonEventsHandlers(method)}
      >
        {icon}
      </div>
    );
  };

  return (
    <Input
      {...omit(rest, ['allowClear'])}
      {...inputEventHandlers}
      style={style}
      className={cs(
        prefixCls,
        `${prefixCls}-mode-${mode}`,
        `${prefixCls}-size-${mergedSize}`,
        {
          [`${prefixCls}-readonly`]: readOnly,
          [`${prefixCls}-illegal-value`]: !isEmptyValue && isOutOfRange,
        },
        className
      )}
      ref={refInput}
      size={mergedSize}
      error={error}
      disabled={disabled}
      readOnly={readOnly}
      value={getDisplayInputValue()}
      placeholder={placeholder}
      prefix={prefix && <div className={`${prefixCls}-prefix`}>{prefix}</div>}
      suffix={
        <>
          {shouldRenderLayer && (
            <div className={`${prefixCls}-step-layer`}>
              {renderStepButton('plus', icons && icons.up ? icons.up : <IconUp />)}
              {renderStepButton('minus', icons && icons.down ? icons.down : <IconDown />)}
            </div>
          )}
          {suffix && <div className={`${prefixCls}-suffix`}>{suffix}</div>}
        </>
      }
      addBefore={
        shouldRenderButton &&
        renderStepButton('minus', icons && icons.minus ? icons.minus : <IconMinus />)
      }
      addAfter={
        shouldRenderButton &&
        renderStepButton('plus', icons && icons.plus ? icons.plus : <IconPlus />)
      }
    />
  );
}

const InputNumberComponent = React.forwardRef<RefInputType, InputNumberProps>(InputNumber);

InputNumberComponent.displayName = 'InputNumber';

export default InputNumberComponent;

export { InputNumberProps };
