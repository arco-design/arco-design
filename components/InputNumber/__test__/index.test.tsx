import React, { useState } from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import InputNumber from '..';

mountTest(InputNumber);
componentConfigTest(InputNumber, 'InputNumber');

function mountInputNumber(component: React.ReactElement) {
  return mount(component);
}

describe('InputNumber ', () => {
  it('init value correctly', () => {
    const wrapper = mountInputNumber(<InputNumber defaultValue={8} min={0} max={15} />);
    const input = wrapper.find('.arco-input');

    expect(+input.prop('value')).toBe(8);
    input.simulate('change', { target: { value: '12' } });
    expect(+wrapper.find('.arco-input').prop('value')).toBe(12);
    input.simulate('change', { target: { value: '123' } });
    expect(+wrapper.find('.arco-input').prop('value')).toBe(123);
    input.simulate('blur');
    expect(+wrapper.find('.arco-input').prop('value')).toBe(15);
    input.simulate('change', { target: { value: '-123' } });
    expect(+wrapper.find('.arco-input').prop('value')).toBe(-123);
    input.simulate('blur');
    expect(+wrapper.find('.arco-input').prop('value')).toBe(0);
  });

  it('init value with empty string correctly', () => {
    const wrapper = mountInputNumber(<InputNumber value="" />);
    expect(wrapper.find('.arco-input').prop('value')).toBe('');
  });

  it('init value with string correctly', () => {
    const wrapper = mountInputNumber(<InputNumber value="8.0000" precision={2} />);
    expect(wrapper.find('.arco-input').prop('value')).toBe('8.00');
  });

  it('value control mode', () => {
    const Demo = () => {
      const [value, setValue] = useState(0);
      return (
        <div>
          <button id="clear" onClick={() => setValue(undefined)}>
            clear
          </button>
          <InputNumber value={value} min={10} />;
        </div>
      );
    };
    const wrapper = mountInputNumber(<Demo />);
    expect(+wrapper.find('.arco-input').prop('value')).toBe(0);
    wrapper.find('#clear').at(0).simulate('click');
    expect(wrapper.find('.arco-input').prop('value')).toBe('');
  });

  it('plus correctly', () => {
    const wrapper = mountInputNumber(<InputNumber min={0} max={15} />);
    const plusButton = wrapper.find('.arco-input-number-step-button').at(0);

    expect(wrapper.find('.arco-input').prop('value')).toBe('');

    for (let i = 0; i <= 20; i++) {
      plusButton.simulate('mousedown');
      plusButton.simulate('mouseup');
      if (i < 15) {
        expect(+wrapper.find('.arco-input').prop('value')).toBe(i);
      } else {
        expect(+wrapper.find('.arco-input').prop('value')).toBe(15);
      }
    }
  });

  it('minus correctly', () => {
    const wrapper = mountInputNumber(<InputNumber min={0} max={15} />);
    const buttons = wrapper.find('.arco-input-number-step-button');
    const plusButton = buttons.at(0);
    const minusButton = buttons.at(1);

    plusButton.simulate('mousedown');
    plusButton.simulate('mouseup');
    expect(+wrapper.find('input').prop('value')).toBe(0);

    for (let i = 0; i < 5; i++) {
      minusButton.simulate('mousedown');
      plusButton.simulate('mouseup');
      expect(+wrapper.find('input').prop('value')).toBe(0);
    }
  });

  it('keyboard plus correctly', () => {
    const wrapper = mountInputNumber(<InputNumber min={0} max={15} />);
    const input = wrapper.find('.arco-input');
    input.simulate('keydown', {
      key: 'ArrowUp',
    });

    expect(+wrapper.find('input').prop('value')).toBe(0);
    for (let i = 0; i < 5; i++) {
      input.simulate('keydown', {
        key: 'ArrowUp',
      });
      expect(+wrapper.find('input').prop('value')).toBe(i + 1);
    }
  });

  it('keyboard minus correctly', () => {
    const wrapper = mountInputNumber(<InputNumber defaultValue={6} min={0} max={15} />);
    const input = wrapper.find('.arco-input');

    input.simulate('keydown', {
      key: 'ArrowDown',
    });
    expect(+wrapper.find('input').prop('value')).toBe(5);

    for (let i = 5; i > 0; i--) {
      input.simulate('keydown', {
        key: 'ArrowDown',
      });
      expect(+wrapper.find('input').prop('value')).toBe(i - 1);
    }
  });

  it('step precision correctly', () => {
    const wrapper = mountInputNumber(<InputNumber defaultValue={6} min={0} max={15} step={0.01} />);
    expect(+wrapper.find('input').prop('value')).toBe(6);
  });

  it('props precision correctly', () => {
    const wrapper = mountInputNumber(
      <InputNumber defaultValue={6} min={0} max={15} step={0.01} precision={3} />
    );
    expect(+wrapper.find('input').prop('value')).toBe(6);
  });

  describe('formatter', () => {
    it('formatter on default', () => {
      const wrapper = mountInputNumber(
        <InputNumber defaultValue={6} min={0} max={15} formatter={(num) => `${num}%`} />
      );
      expect(wrapper.find('input').prop('value')).toBe('6%');
    });

    it('formatter on mousedown', () => {
      const wrapper = mountInputNumber(
        <InputNumber defaultValue={6} min={0} max={15} formatter={(num) => `${num}%`} />
      );
      const buttons = wrapper.find('.arco-input-number-step-button');
      const plusButton = buttons.at(0);
      const minusButton = buttons.at(1);

      plusButton.simulate('mousedown');
      plusButton.simulate('mouseup');
      expect(wrapper.find('input').prop('value')).toBe('7%');

      minusButton.simulate('mousedown');
      minusButton.simulate('mouseup');
      expect(wrapper.find('input').prop('value')).toBe('6%');
    });

    it('formatter on keydown', () => {
      const wrapper = mountInputNumber(
        <InputNumber defaultValue={6} min={0} max={15} formatter={(num) => `${num}%`} />
      );
      const input = wrapper.find('.arco-input');

      input.simulate('keydown', {
        key: 'ArrowDown',
      });
      expect(wrapper.find('input').prop('value')).toBe('5%');

      for (let i = 5; i > 0; i--) {
        input.simulate('keydown', {
          key: 'ArrowDown',
        });
        expect(wrapper.find('input').prop('value')).toBe(`${i - 1}%`);
      }
      expect(wrapper.find('input').prop('value')).toBe('0%');

      for (let i = 0; i < 5; i++) {
        input.simulate('keydown', {
          key: 'ArrowUp',
        });
        expect(wrapper.find('input').prop('value')).toBe(`${i + 1}%`);
      }
    });

    it('formatter on input', () => {
      let changeValue = 6;
      const wrapper = mountInputNumber(
        <InputNumber
          defaultValue={6}
          min={0}
          max={15}
          formatter={(num) => `${num}%`}
          onChange={(val) => (changeValue = val)}
        />
      );
      const input = wrapper.find('.arco-input');

      input.simulate('focus');
      input.simulate('change', { target: { value: '12' } });
      expect(wrapper.find('input').prop('value')).toBe('12%');
      expect(changeValue).toBe(12);
      input.simulate('blur');
      expect(wrapper.find('input').prop('value')).toBe('12%');
    });

    it('formatter and parser', () => {
      let changeValue = 0;
      const wrapper = mountInputNumber(
        <InputNumber
          min={0}
          max={1000000000}
          onChange={(val) => (changeValue = val)}
          parser={(value) => Number(value) / 100}
          formatter={(value) => `${Number(value) * 100}`}
          step={0.1}
        />
      );
      const input = wrapper.find('.arco-input');
      input.simulate('focus');
      input.simulate('change', { target: { value: '12' } });
      expect(wrapper.find('input').prop('value')).toBe('12');
      expect(changeValue).toBe(0.12);
      input.simulate('blur');
      expect(wrapper.find('input').prop('value')).toBe('12');
      expect(changeValue).toBe(0.12);
    });
  });

  describe('blur and focus', () => {
    it('onBlur and onFocus', () => {
      const onBlur = jest.fn();
      const onFocus = jest.fn();
      const wrapper = mountInputNumber(
        <InputNumber onBlur={onBlur} onFocus={onFocus} defaultValue={7} min={0} max={15} />
      );
      const input = wrapper.find('.arco-input');

      input.simulate('focus');
      expect(onFocus.mock.calls.length).toBe(1);
      input.simulate('blur');
      expect(onBlur.mock.calls.length).toBe(1);
    });
  });

  it('max and min works fine', () => {
    const Demo = () => {
      const [value, setValue] = useState(5);
      return (
        <div>
          <button onClick={() => setValue(value + 10)}>Add</button>
          <InputNumber min={10} max={20} value={value} onChange={setValue} />
        </div>
      );
    };
    const wrapper = mount(<Demo />);

    expect(wrapper.find('input').prop('value')).toBe('5');

    wrapper.find('button').simulate('click');
    expect(wrapper.find('input').prop('value')).toBe('15');

    wrapper.find('button').simulate('click');
    expect(wrapper.find('input').prop('value')).toBe('25');

    wrapper.find('input').simulate('focus');
    wrapper.find('input').simulate('blur');
    expect(wrapper.find('input').prop('value')).toBe('20');
  });
});
