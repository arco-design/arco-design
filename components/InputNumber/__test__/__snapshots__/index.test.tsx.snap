// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.InputNumber default 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-input-number arco-input-number-mode-embed arco-input-number-size-default"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        aria-valuemax="Infinity"
        aria-valuemin="-Infinity"
        class="arco-input arco-input-size-default"
        role="spinbutton"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <div
          class="arco-input-number-step-layer"
        >
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-up"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 30.557 24.043 15 8.487 30.557"
              />
            </svg>
          </div>
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-down"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 17.443 24.043 33 8.487 17.443"
              />
            </svg>
          </div>
        </div>
      </span>
    </span>
  </span>
</div>
`;

exports[`ConfigProvider componentConfig.InputNumber set className globally 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-input-number arco-input-number-mode-embed arco-input-number-size-default global-default-class"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        aria-valuemax="Infinity"
        aria-valuemin="-Infinity"
        class="arco-input arco-input-size-default"
        role="spinbutton"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <div
          class="arco-input-number-step-layer"
        >
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-up"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 30.557 24.043 15 8.487 30.557"
              />
            </svg>
          </div>
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-down"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 17.443 24.043 33 8.487 17.443"
              />
            </svg>
          </div>
        </div>
      </span>
    </span>
  </span>
</div>
`;

exports[`ConfigProvider componentConfig.InputNumber set className globally and component set className itself. 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-input-number arco-input-number-mode-embed arco-input-number-size-default component-class"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        aria-valuemax="Infinity"
        aria-valuemin="-Infinity"
        class="arco-input arco-input-size-default"
        role="spinbutton"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <div
          class="arco-input-number-step-layer"
        >
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-up"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 30.557 24.043 15 8.487 30.557"
              />
            </svg>
          </div>
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-down"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 17.443 24.043 33 8.487 17.443"
              />
            </svg>
          </div>
        </div>
      </span>
    </span>
  </span>
</div>
`;

exports[`ConfigProvider componentConfig.InputNumber set data-* property 1`] = `
<div
  class="arco-input-group-wrapper arco-input-group-wrapper-default arco-input-has-suffix arco-input-number arco-input-number-mode-embed arco-input-number-size-default"
>
  <span
    class="arco-input-group"
  >
    <span
      class="arco-input-inner-wrapper arco-input-inner-wrapper-default"
    >
      <input
        aria-valuemax="Infinity"
        aria-valuemin="-Infinity"
        class="arco-input arco-input-size-default"
        data-test-name="InputNumber"
        role="spinbutton"
        value=""
      />
      <span
        class="arco-input-group-suffix"
      >
        <div
          class="arco-input-number-step-layer"
        >
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-up"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 30.557 24.043 15 8.487 30.557"
              />
            </svg>
          </div>
          <div
            class="arco-input-number-step-button"
          >
            <svg
              aria-hidden="true"
              class="arco-icon arco-icon-down"
              fill="none"
              focusable="false"
              stroke="currentColor"
              stroke-width="4"
              viewBox="0 0 48 48"
            >
              <path
                d="M39.6 17.443 24.043 33 8.487 17.443"
              />
            </svg>
          </div>
        </div>
      </span>
    </span>
  </span>
</div>
`;
